[BUG] Serialization error during peer-forwarding

**Describe the bug**

Peer forwarding is unable to serialize events.

```
2024-01-17T23:51:34.293 [test-pipeline-processor-worker-1-thread-1] WARN  org.opensearch.dataprepper.peerforwarder.RemotePeerForwarder - Unable to submit request for forwarding, processing locally.
java.lang.RuntimeException: java.io.NotSerializableException: java.lang.ref.WeakReference
  at org.opensearch.dataprepper.peerforwarder.client.PeerForwarderClient.getSerializedJsonBytes(PeerForwarderClient.java:87) ~[data-prepper-core-2.6.1.jar:?]
  at org.opensearch.dataprepper.peerforwarder.client.PeerForwarderClient.serializeRecordsAndSendHttpRequest(PeerForwarderClient.java:71) ~[data-prepper-core-2.6.1.jar:?]
  at org.opensearch.dataprepper.peerforwarder.RemotePeerForwarder.forwardRecordsForIp(RemotePeerForwarder.java:267) ~[data-prepper-core-2.6.1.jar:?]
Caused by: java.io.NotSerializableException: java.lang.ref.WeakReference
  at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1175) ~[?:?]
  at java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1543) ~[?:?]
  at java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1500) ~[?:?]
  at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1423) ~[?:?]
  at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1169) ~[?:?]
  at java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1543) ~[?:?]
  at java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1500) ~[?:?]
  ... 14 more
```

**To Reproduce**

1. Configure Data Prepper to use peer-forwarding
2. Create a pipeline with an `aggregate` processor.
3. Run multiple Data Prepper instances
4. Inject data that will need to route across peers

 
**Expected behavior**

One Data Prepper node should be able to serialize and send the data to the other Data Prepper node.

**Environment (please complete the following information):**
 - Data Prepper: 2.6.1

--- a/data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/JacksonEvent.java
+++ b/data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/JacksonEvent.java
@@ -23,6 +23,7 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
+import java.io.ObjectInputStream;
 import java.time.Instant;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -68,7 +69,7 @@ public class JacksonEvent implements Event {
 
     private final EventMetadata eventMetadata;
 
-    private EventHandle eventHandle;
+    private transient EventHandle eventHandle;
 
     private final JsonNode jsonNode;
 
@@ -657,7 +658,17 @@ public class JacksonEvent implements Event {
 
             }
         }
+    }
 
-
+    /**
+     * Provides custom Java object deserialization.
+     *
+     * @param objectInputStream The {@link ObjectInputStream} to deserialize from
+     * @throws IOException if an I/O error occurs
+     * @throws ClassNotFoundException if the class of a serialized object could not be found
+     */
+    private void readObject(final ObjectInputStream objectInputStream) throws IOException, ClassNotFoundException {
+        objectInputStream.defaultReadObject();
+        this.eventHandle = new DefaultEventHandle(eventMetadata.getTimeReceived());
     }
 }

--- a/data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEvent_JavaSerializationTest.java
+++ b/data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEvent_JavaSerializationTest.java
@@ -1,0 +1,87 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.model.event;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSet;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.Map;
+import java.util.UUID;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.nullValue;
+import static org.mockito.Mockito.mock;
+
+class JacksonEvent_JavaSerializationTest {
+
+    private ObjectOutputStream objectOutputStream;
+    private ByteArrayOutputStream byteArrayOutputStream;
+
+    @BeforeEach
+    void setUp() throws IOException {
+        byteArrayOutputStream = new ByteArrayOutputStream(1000);
+        objectOutputStream = new ObjectOutputStream(byteArrayOutputStream);
+    }
+
+    private JacksonEvent createObjectUnderTest() {
+        return JacksonEvent.builder()
+                .withEventType("TEST")
+                .withData(Map.of(UUID.randomUUID().toString(), UUID.randomUUID().toString()))
+                .build();
+    }
+
+    @Test
+    void serialize_without_acknowledgementSet_includes_data() throws IOException, ClassNotFoundException {
+        final JacksonEvent objectUnderTest = createObjectUnderTest();
+
+        final Object deserializedObject = serializeAndDeserialize(objectUnderTest);
+
+        assertThat(deserializedObject, instanceOf(JacksonEvent.class));
+        final JacksonEvent deserializedEvent = (JacksonEvent) deserializedObject;
+
+        assertThat(deserializedEvent.toMap(), equalTo(objectUnderTest.toMap()));
+        assertThat(deserializedEvent.getMetadata(), equalTo(objectUnderTest.getMetadata()));
+
+        assertThat(deserializedEvent.getEventHandle(), instanceOf(InternalEventHandle.class));
+        assertThat(((InternalEventHandle) deserializedEvent.getEventHandle()).getAcknowledgementSet(), nullValue());
+        assertThat(deserializedEvent.getEventHandle().getInternalOriginationTime(), equalTo(objectUnderTest.getMetadata().getTimeReceived()));
+
+    }
+
+    @Test
+    void serialize_with_acknowledgementSet_does_not_include_old_acknowledgement_set() throws IOException, ClassNotFoundException {
+        final JacksonEvent objectUnderTest = createObjectUnderTest();
+        final InternalEventHandle internalEventHandle = (InternalEventHandle) objectUnderTest.getEventHandle();
+        internalEventHandle.setAcknowledgementSet(mock(AcknowledgementSet.class));
+
+        final Object deserializedObject = serializeAndDeserialize(objectUnderTest);
+
+        assertThat(deserializedObject, instanceOf(JacksonEvent.class));
+        final JacksonEvent deserializedEvent = (JacksonEvent) deserializedObject;
+
+        assertThat(deserializedEvent.toMap(), equalTo(objectUnderTest.toMap()));
+        assertThat(deserializedEvent.getMetadata(), equalTo(objectUnderTest.getMetadata()));
+
+        assertThat(deserializedEvent.getEventHandle(), instanceOf(InternalEventHandle.class));
+        assertThat(((InternalEventHandle) deserializedEvent.getEventHandle()).getAcknowledgementSet(), nullValue());
+        assertThat(deserializedEvent.getEventHandle().getInternalOriginationTime(), equalTo(objectUnderTest.getMetadata().getTimeReceived()));
+    }
+
+    private Object serializeAndDeserialize(final JacksonEvent objectUnderTest) throws IOException, ClassNotFoundException {
+        objectOutputStream.writeObject(objectUnderTest);
+        final ObjectInputStream objectInputStream = new ObjectInputStream(new ByteArrayInputStream(byteArrayOutputStream.toByteArray()));
+        return objectInputStream.readObject();
+    }
+
+}

--- a/data-prepper-core/src/test/java/org/opensearch/dataprepper/peerforwarder/codec/JavaPeerForwarderCodecTest.java
+++ b/data-prepper-core/src/test/java/org/opensearch/dataprepper/peerforwarder/codec/JavaPeerForwarderCodecTest.java
@@ -2,7 +2,9 @@ package org.opensearch.dataprepper.peerforwarder.codec;
 
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSet;
 import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.event.InternalEventHandle;
 import org.opensearch.dataprepper.model.event.JacksonEvent;
 import org.opensearch.dataprepper.model.log.JacksonLog;
 import org.opensearch.dataprepper.peerforwarder.model.PeerForwardingEvents;
@@ -54,6 +56,24 @@ class JavaPeerForwarderCodecTest {
         when(objectInputFilter.checkInput(any(ObjectInputFilter.FilterInfo.class))).thenReturn(ObjectInputFilter.Status.ALLOWED);
 
         final PeerForwardingEvents inputEvents = generatePeerForwardingEvents(2);
+        final byte[] bytes = createObjectUnderTest().serialize(inputEvents);
+        final PeerForwardingEvents outputEvents = createObjectUnderTest().deserialize(bytes);
+        assertThat(outputEvents.getDestinationPipelineName(), equalTo(inputEvents.getDestinationPipelineName()));
+        assertThat(outputEvents.getDestinationPluginId(), equalTo(inputEvents.getDestinationPluginId()));
+        assertThat(outputEvents.getEvents().size(), equalTo(inputEvents.getEvents().size()));
+
+        verify(objectInputFilter, atLeast(1)).checkInput(any(ObjectInputFilter.FilterInfo.class));
+    }
+
+    @Test
+    void testCodec_with_acknowledgementSet() throws IOException, ClassNotFoundException {
+        when(objectInputFilter.checkInput(any(ObjectInputFilter.FilterInfo.class))).thenReturn(ObjectInputFilter.Status.ALLOWED);
+
+        final PeerForwardingEvents inputEvents = generatePeerForwardingEvents(2);
+        inputEvents.getEvents().stream()
+                .map(Event::getEventHandle)
+                .map(handle -> (InternalEventHandle)handle)
+                .forEach(handle -> handle.setAcknowledgementSet(mock(AcknowledgementSet.class)));
         final byte[] bytes = createObjectUnderTest().serialize(inputEvents);
         final PeerForwardingEvents outputEvents = createObjectUnderTest().deserialize(bytes);
         assertThat(outputEvents.getDestinationPipelineName(), equalTo(inputEvents.getDestinationPipelineName()));

